### ISSUE TITLE
Dumping large files is slow

### ISSUE THREAD

**Issue by merrygoat** [LightForm-group/xrdfit#30]
Dumping large experiments seems to be slow. This may be a consequence of the heavily nested objects and dill is just slow, perhaps it relates to the file compression as it is dumped. This could easily be profiled.

**Comment by merrygoat** [comment]
Some benchmarking for ~ 60 analysis frames of 10 peaks:

- uncompressed - 10 s dump, 2 s load - 22 Mb file
- lz4 compression - 15 s dump, 2 s load - 11 Mb file
- bz4 compression - 15 s dump, 3 s load - 6 Mb file

The code is already using the bz4 and I think this is the best choice. The compression has some effect on speed but is worth it with a 70% decrease in file size. The main problem is saving the lmfit Result objects which are large and heavily nested.

The next thing to look at might be serialising everything and dumping as json - the model result has a "dumps" method.

**Comment by merrygoat** [comment]
JSON serialisation uncompressed - 5 s dump - 7 Mb
JSON serialisation + bz4 compression - 5.5s - 0.9 Mb

OK we have a winner here! Have to write the unpacking methods now. It might be that unpacking is now slow since we have lost the native binary data of pickle but will have to see.

**Comment by merrygoat** [comment]
This is hard because the reconstituted object doesn't have the same structure as the parameters used to instantiate PeakParams. The solution is to change the signature of PeakParams `__init__` to match the dumped data and pre process user input with a Factory function to call the `__init__`.

**Comment by merrygoat** [comment]
Now at: 
JSON serialisation + bz4 compression - 9 s dump, 5 s load - 5 Mb

My initial test missed saving the spectral data which is why it was faster and the file smaller. This is a good result in a way since it shows that my implementation isn't doing any better than dill which I wouldn't really expect it to do since it is doing mostly the same thing.

This implementation also comes with a heavy cost in that I have written a basic dill into my code and so have to maintain it!

This raises the interesting point that it is the spectral data which takes up a lot of the room/time in saving/loading the data. This is because every one of the cakes is loaded - even those that aren't fitted. I would have thought the lmfit ModelResults would have been the thing taking the space but of course, they only contain the data from the fitted cakes. One could choose not to save and reload this data but that would likely defeat the point of saving the whole object.

**Comment by merrygoat** [comment]
bz4 compression, remove raw data - 13 s dump, 2.5 s load - 2 Mb file
bz4 compression, remove lmfit ModelResult - 4s dump, 0.5s load, 4 Mb file

Profiling this is a real mess - I keep getting confused. My first guess was actually the right one and the profiling mislead me. It is not really data size or compression that is slow, it is the heavily nested lmfit ModelResults, packing and unpacking them takes the time.

There is no real fix to this aside from don't use lmfit (which is an interesting consideration for increasing fitting performance too) which is way outside the scope of this issue or even this project until it is being heavily utilised.

In conclusion, slowness is a feature not a bug so no fix.
